### DISCUSSION TITLE
how to  improve the efficiency of HPC process?

### DISCUSSION THREAD

**Discussion by Sun-Liu-Yang** [idaholab/moose#29088] [Q&A Tools]
I am trying to use MPI functionality for large-scale parallel computing on a supercomputer cluster. The number of solving units is in the tens of millions, and the number of degrees of freedom of nodes is in the billions.
When I was solving a problem with millions of degrees of freedom, I found that the solving efficiency was already very slow. The official suggested a single core degree of freedom of 20000, but currently my single core solving degree of freedom is only about 10000.
What aspects should I improve the calculation settings to significantly increase computational efficiency and meet the demand for solving billions of degrees of freedom?

**Comment by GiudGiud** [comment]
Hello

This is a complex problem and the simplest step to take is simply to profile a shorter run to see which step in your simulation is not scaling to tens of millions of cores. The limiting step is likely specific to your problem and to your solver setup.
https://mooseframework.inl.gov/application_development/profiling.html

**Reply by Sun-Liu-Yang** [reply]
how to reduce the memory occupition？
when I using the MPI funtion, the  memory occupition of the CPU is very, about 30G/CPU  on a supercomputer cluster. But the memory of the cluster is not enough and the effecifency is slow. 
the elements number of my model is about 30,000,000, and the freedom is almost 100,000,000.
the list is my input file of the model.
thank you very much!
```
[Mesh]
  type = FileMesh
  file ='2024-11-22.inp'
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Variables]
  [disp_x]
  []

  [disp_y]
  []

  [disp_z]
  []
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = FINITE
        add_variables = false
        generate_output = 'stress_xx 
                                    stress_yy 
                            stress_zz 
                           min_principal_stress max_principal_stress
                           volumetric_strain vonmises_stress'
      []
    []
  []
[]


[Functions]

  [axial_pressure_1]
    type = PiecewiseLinear
    x = '0  1  2'
    y = '0  1e6  2e6'
  []
 

  [confining_pressure]
    type = PiecewiseLinear
    x = '0  1   1.5  2'
    y = '0  1e6 1e6 1e6'
  []
[]

[BCs]

  [Pressure]

    [load1]
      boundary = 'up_to_force'
      function = 'axial_pressure_1'        # Pa
    []

    [load2]
      boundary = 'side_to_confing'
      function = 'confining_pressure'   # Pa
    []

  []

  [anchor_x]
    type = DirichletBC
    variable = disp_x
    boundary = '0'
    value = 0.0
  []

  [anchor_y]
    type = DirichletBC
    variable = disp_y
    boundary = '0'
    value = 0.0
  []

  [anchor_z]
    type = DirichletBC
    variable = disp_z
    boundary = '0'
    value = 0.0
  []

[]

[Materials]

  [elasticity_tensor_soil]
    type = ComputeIsotropicElasticityTensor
    bulk_modulus = 50e6
    poissons_ratio = 0.25
  []

  [stress]
    type = ComputeFiniteStrainElasticStress
  []

  [density]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 1800 # kg/m^3
  []

[]




[Executioner]

  type = Transient
  solve_type = JFNK
  l_max_its = 100000
  nl_max_its = 100
  l_tol = 1E-6
  nl_abs_tol = 1E-4
  start_time = 0.0
  end_time = 0.02
  dt = 0.01
  dtmin = 0.001

[]

[Outputs]
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
Reducing the number of outputs will reduce the number of fields so i would advise doing that
```
        generate_output = 'stress_xx 
                                    stress_yy 
                            stress_zz 
                           min_principal_stress max_principal_stress
                           volumetric_strain vonmises_stress'
```

Then are you getting good convergence with JFNK? That's fairly uncommon. 
I would recommend you switch to PJFNK or Newton and use Hypre boomeramg which should scale well for solid mechanics

**Reply by Sun-Liu-Yang** [reply]
Hello there is a new problem I want to request.
My test model include over 30,000,000 elements and the total Dofs is over 100,000,000.
I used the mesh spliting function pre-spliting the mesh into 3072 partitons ,and resubmited  the job using the mesh spilitting configuration .cpr file. But it take several hours to prepare the mesh and cannot continue.    

 Preparing Mesh state resisting several hours and cannot continue.
 
mesh splitting common
resubmit commond : mpiexec -n 3072 ./solid_mechanics-opt-intel_2024 -i triaxial/123.i --use-split --split-file 2024-11-22.cpr
mesh spilitting commond: mpiexec -n 3072  ./solid_mechanics-opt-intel_2024 -i triaxial/123.i   --split-mesh 3072  --split-file 2024-11-22.cpr


calculating state ：


[36m
*** Info ***
SolidMechanics Action: selecting 'total small strain' formulation. Use `incremental = true` to select 'incremental small strain' instead.[39m
Currently Setting Up
  Setting Mesh
    Reading Mesh......                                                                   [[33m 43.21 s[39m] [[33m  361 MB[39m]
  Finished Setting Mesh                                                                  [[33m 43.29 s[39m] [[33m  361 MB[39m]
  Setting Up Undisplaced Mesh
    Preparing Mesh........................................................

**Reply by GiudGiud** [reply]
Hello

Using many parallel processes to split it actually a bad idea, because it's replicated during that phase.
Try this:
```
./solid_mechanics-opt-intel_2024 -i triaxial/123.i --split-mesh 3072 --split-file 2024-11-22.cpr
```
and if it does not work, use as few mpi processes as possible

**Reply by Sun-Liu-Yang** [reply]
Thanks very much your help on this problem, but there is another problem disturbing me.
the case is that following: if the model incldues 20000 elements, about 6000dofs.
**case1:**  the model construcde by moose's mesh generator (type = meshgenerate/meshgenerator)
I use the pre-split commond to part the model into 8 partitions and sovle the problems  using 2 nodes and 8 processer  on HPC. In this case ,it is very effecient to read and solve the problems.
```
 [Mesh]
    type = GeneratedMesh
    dim = 3
    nx = 20
    ny = 20
    nz = 50
    xmax = 20
    xmin = 0
    ymax = 20
    ymin = 0
    zmax = 50
    zmin = 0
    [Partitioner]
      type = PetscExternalPartitioner
      part_package = parmetis
   []  
  [Partitioner]
      type = PetscExternalPartitioner
      part_package = parmetis
  []  
[]
..............
[Executioner]

  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-pc_type  -sub_pc_type'
  petsc_options_value =  'bjacobi  ilu'
  l_max_its = 100000
  nl_max_its = 100
  l_tol = 1E-6
  nl_abs_tol = 1E-4
  start_time = 0.0
  end_time = 0.1
  dt = 0.1
  dtmin = 0.01

[]

[Outputs]
  [out_Nemesis]
    type = Nemesis
  []
  [perf_out]
    type = PerfGraphOutput
    execute_on = 'initial final'
    level = 3
  []
[]
```
**mesh pre-split :** mpiexec -n 2  ./solid_mechanics-opt-intel_2024 -i 20250108.i  --split-mesh 8 --split-file 2025cir.cpr
**job submission:** mpiexec -n 100 ./solid_mechanics-opt-intel_2024 -i 20250108.i --use-split --split-file 2025cir.cpr

**case2 :**  the model construcde by exported  mesh file (type = meshfile, file = '2345.inp')
![image](https://github.com/user-attachments/assets/e10bcf42-1f1d-4e60-9d2f-3d863877ac1c)
Firstly,  I also use the pre-split commond to part the model into 8 partitions. But the troubles comes following.
If the problem is solved using 8cpus in only 1 nodes, it is also very effecient to read the pre-split mesh and solve the problems.
The problem is **If the  problem is solved using 8cpus and 2 nodes (4cpus/1node), it will take very long time in the preparing mesh state and on processing.**
```
 [Mesh]
    type = FileMesh
    file ='20250108.inp' 
  [Partitioner]
      type = PetscExternalPartitioner
      part_package = parmetis
  []  
[]
..............
[Executioner]

  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-pc_type  -sub_pc_type'
  petsc_options_value =  'bjacobi  ilu'
  l_max_its = 100000
  nl_max_its = 100
  l_tol = 1E-6
  nl_abs_tol = 1E-4
  start_time = 0.0
  end_time = 0.1
  dt = 0.1
  dtmin = 0.01
[]

[Outputs]
  [out_Nemesis]
    type = Nemesis
  []
  [perf_out]
    type = PerfGraphOutput
    execute_on = 'initial final'
    level = 3
  []
[]
```
**mesh pre-split :** mpiexec -n 2  ./solid_mechanics-opt-intel_2024 -i 20250108.i  --split-mesh 8 --split-file 2025cir.cpr
**job submission:** mpiexec -n 100 ./solid_mechanics-opt-intel_2024 -i 20250108.i --use-split --split-file 2025cir.cpr

**Reply by GiudGiud** [reply]
hello

this `mpiexec -n 100 ./solid_mechanics-opt-intel_2024 -i 20250108.i --use-split --split-file 2025cir.cpr` is wanting to use 100 partitions not 8. So my guess is that you are not actually using a split mesh right now, since the split mesh has been generated for 8 partitions

**Reply by Sun-Liu-Yang** [reply]
when run the problem using 2nodes, the processing state is like this 
![image](https://github.com/user-attachments/assets/6f556f7a-5114-494f-978c-0e6c880ea4f0)
and when run the proble using 1 node ,the processing state is like this 
![image](https://github.com/user-attachments/assets/01f0c4e7-f8fb-46aa-8c54-465f8a2ee8e3)

**Reply by Sun-Liu-Yang** [reply]
sorry  I spelling wrong the commond in the problem description, but the problem is that when dealing the problem with exported mesh model  on more than 1 HPC nodes, it cannot proceeding.

**Reply by GiudGiud** [reply]
what commands are you running then?
inter-node communication being slower than intra-node is expected. But not by a lot either

**Reply by Sun-Liu-Yang** [reply]
In fact, when using the moose's  meshgenerator constructing the model, even the elements over 200milllion, it can also dealing very fast on 4 nodes and 200 cpus.
but when the model mesh from the abaqus(the .inp file ), only in one node can dealing fast.

The command is right when I submit the jobs.
For example:
**pre-spit command**
```
#SBATCH --job-name=moose
#SBATCH -p intel_expr
#SBATCH --nodes=2 
#SBATCH --ntasks=2
#SBATCH -c 1 
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --output=moose-%j.out
#SBATCH --mem=500G
#SBATCH -o 20250108.log
module load intel/oneapi/2024.2.0
# export OMP_NUM_THREADS=1
mpiexec -n 2 ./solid_mechanics-opt-intel_2024  -i  triaxial/2345.i --split-mesh  60
```
**Job submitting command**
```
#!/bin/sh
#SBATCH --job-name=moose
#SBATCH -p intel_expr
#SBATCH --nodes=2 
#SBATCH --ntasks=60
#SBATCH -c 1 
#SBATCH --ntasks-per-node=30
#SBATCH --cpus-per-task=1
#SBATCH --output=moose-%j.out
#SBATCH --mem=500G
#SBATCH -o 20250108.log
module load intel/oneapi/2024.2.0
# export OMP_NUM_THREADS=1
mpiexec -n 60 ./solid_mechanics-opt-intel_2024  -i  triaxial/2345.i --use-split --split-file 2024-11-14-55.cpa.gz
```

**Reply by GiudGiud** [reply]
> but when the model mesh from the abaqus(the .inp file ), only in one node can dealing fast.

do you mean with the split mesh? Reading an inp file directly would be slow since it's replicated.

There is an other way to run split meshes that might work better. 
Remove --use-split and instead use a FileMeshGenerator with the file set to `2024-11-14-55.cpa.gz`.
And add `--distributed-mesh` to the command line

**Reply by Sun-Liu-Yang** [reply]
I have try this method to solve the problem ,it seems not work inter-node.

![image](https://github.com/user-attachments/assets/8fccdfd8-9291-4d46-9bf8-daf29af45949)
![image](https://github.com/user-attachments/assets/1cf604c1-e954-421d-ba0c-1c2e37e37aca)
![image](https://github.com/user-attachments/assets/c1ef1cc8-c69a-4f5e-b48e-e90396a58a41)

**Reply by Sun-Liu-Yang** [reply]
this figure is the same model solved on 1nodes with 60cpus.
![image](https://github.com/user-attachments/assets/72d39754-b0f7-4232-8cf0-8b091793a808)

**Reply by GiudGiud** [reply]
I wonder if the partitioner is the problem here. 

If it returned two different partitions on both runs, it would fail. 

Can you try commenting it out? Parmetis is the default anyway, you are simply changing how we access parmetis by specifying this PetscPartitioner

**Reply by Sun-Liu-Yang** [reply]
you means when splitting the mesh  and submit the job，commenting the  [Partitioner] block out? 
or
 only  commenting the  [Partitioner] block out when submitting the job?

**Comment by Sun-Liu-Yang** [comment]
Just now, I comment the [Partitioner] block out both the two step, but it dont woke.

![image](https://github.com/user-attachments/assets/bee016e8-0949-46ef-852a-d54a8a2abbd4)

![image](https://github.com/user-attachments/assets/b3a597db-2aea-4d7d-99b7-298dd94df0fa)

![image](https://github.com/user-attachments/assets/2eeebd37-ff77-472d-9ef6-36b9a47a9221)

**Reply by GiudGiud** [reply]
I read the thread again. I think your executable may be hanging as soon as there is internode communication. The mesh preparation is the first step to have such communication.

You can try the same 2 node run without a split mesh to double check. You should see the same issue.
Do runs with just a Cartesian mesh generator solve properly? For them the first communication step will be during the solve

**Reply by Sun-Liu-Yang** [reply]
yes, the problem will be solve properly when the model mesh is constructed by the Moose's mesh geneartor (such as GeneratedMesh object )on more than one nodes.

you means that when dealing with the problem using the external mesh (such as .e, .inp file ), I should take more time for node communicatin until the preparing mesh state finshed ?

**Reply by GiudGiud** [reply]
> yes, the problem will be solve properly when the model mesh is constructed by the Moose's mesh geneartor (such as GeneratedMesh object )on more than one nodes.

Ok that's good to know. So it is the FileMeshGenerator that is slow. 
At this point it would be good to find out why.
You can either use a profiler (make sure to pick a mesh that is not too big so the mesh loading finishes):
https://mooseframework.inl.gov/application_development/profiling.html
or simply use a debugger and interrupt the code while it is "preparing" to see what it is actually is doing
https://mooseframework.inl.gov/application_development/debugging.html

**Reply by Sun-Liu-Yang** [reply]
Thank you very much , I will try to do this .
Whether you have met the similar case when dealing with problems using the external mesh (using the filemesh or filemeshgenerator)? I want to know whether is there some error when installing the moose the my HPC?

**Reply by GiudGiud** [reply]
> Whether you have met the similar case when dealing with problems using the external mesh (using the filemesh or filemeshgenerator)?

Never had this issue. Splitting can be slow if done in parallel but loading splits has always been fine for me

>  I want to know whether is there some error when installing the moose the my HPC?

I dont know if it's that.

**Comment by yfun0815** [comment]
I have a similar issue when preparing mesh... for a long time. It is the issue of your installation of Openmpi of MOOSE. You can contact your hpc.

**Comment by yfun0815** [comment]
Hi @GiudGiud,
I have a 0.9 billion mesh, I want to load it with 1024 or 2048 cores, but both failed because of OOM.
I tried two methods. One is run with --use-split --split-file split.cpr. Another is run with --distributed-mesh and provide split.cpr in input file. The log is:
`[pwx0406@login03] djdjdj_remesh_adjust_kernel_r3_core0]$ cat slurm-3516777.out 
Setting Up..--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
   Current memory used:  4366955356 bytes
   Maximum memory used:  4366955356 bytes
***Memory allocation failed for gk_mcoremalloc: ptr. Requested size: 4294967296 bytes
[qhcn355:1699617] *** Process received signal ***
[qhcn355:1699617] Signal: Aborted (6)
[qhcn355:1699617] Signal code:  (-6)
[qhcn355:1699617] [ 0] /usr/lib64/libpthread.so.0(+0x12cf0)[0x1551fb553cf0]
[qhcn355:1699617] [ 1] /usr/lib64/libpthread.so.0(raise+0x10f)[0x1551fb553b8f]`
Seems the mesh is replicated in every core.

**Reply by GiudGiud** [reply]
hello
 
the split mesh should be generated with very few cores ideally. 
Then once it is split it can be loaded in parallel

**Reply by yfun0815** [reply]
I splited it in only one core and load it with 1024 cores. I have done this with a smaller mesh by 256 cores successfully. But it does not work with this large mesh.

**Reply by yfun0815** [reply]
So, do you mean I need to split the mesh with 1024 cores? Otherwise, I can not load the distributed mesh splited by 1 core.

**Reply by GiudGiud** [reply]
> So, do you mean I need to split the mesh with 1024 cores? Otherwise, I can not load the distributed mesh splited by 1 core.

no your split it using 1 core in 1024 parts, then you load it using `--use-split` with 1024 processes

**Reply by GiudGiud** [reply]
note that the FileMeshGenerator might be easier to use than --use-split

**Reply by yfun0815** [reply]
I noticed memory is the limit for large mesh, so my logic is --use-split will load the distributed mesh. However the memory is still not enough. What do you mean by using FileMeshGenerator?

**Reply by GiudGiud** [reply]
In the [mesh] block you can create a FileMeshGenerator to load a split mesh

**Reply by yfun0815** [reply]
Got it. How to locate the split mesh file and do I still need a --distributed-mesh in the command line?

**Reply by GiudGiud** [reply]
You create the split mesh using your current workflow.
It's for loading it that you can use the FileMeshGenerator

--distributed-mesh can be used when loading the split mesh yes, though I think it's implied
There is also Mesh/parallel_type = distributed you can use instead. 

I recommend using two input files, one for splitting and one for loading the mesh

**Reply by yfun0815** [reply]
Thanks, I have tried your instruction. 
Step 1:
I split the mesh in 1024 parts.
moose -i test1.i --split-mesh 1024 --split-file split.cpr

Step 2:
Another input file is:
[Mesh]
    [mesh]
        type = FileMeshGenerator
        file = split.cpr
    []
    parallel_type = DISTRIBUTED
[]
And the command is:
mpirun -np 1024 moose -i test2.i
Then, the error is:

Setting Up...   Current memory used:  4438531252 bytes
   Maximum memory used:  4438531252 bytes
***Memory allocation failed for gk_mcoremalloc: ptr. Requested size: 4294967296 bytes
   Current memory used:  4439306704 bytes
   Maximum memory used:  4439306704 bytes
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
***Memory allocation failed for gk_mcoremalloc: ptr. Requested size: 4294967296 bytes[qhcn105:3184117] *** Process received signal ***
[qhcn105:3184117] Signal: Aborted (6)
[qhcn105:3184117] Signal code:  (-6)
   Current memory used:  4438541676 bytes
   Maximum memory used:  4438541676 bytes
   
I believe 1024 cores have 8T memory in total, which should be enough. So I suspect the mesh wasn't loaded vis distributed mode.

**Reply by GiudGiud** [reply]
Same results with --distributed-mesh on the command line?

Can you check with 
mpirun -np 1024 hostname
 that the processes are being assigned to different nodes? (Groups of processes on each node are ok)

Are you using slurm? PBS? For the scheduler

**Reply by yfun0815** [reply]
> Same results with --distributed-mesh on the command line?

Yes, both terminate at Setting up...

> Can you check with
mpirun -np 1024 hostname
that the processes are being assigned to different nodes? (Groups of processes on each node are ok)

This is ok. And I can run a smaller model via multiple nodes with a fair speed.

> Are you using slurm? PBS? For the scheduler

Here is my job script:
#!/bin/bash
#SBATCH -p intel
#SBATCH -N 16
#SBATCH --ntasks-per-node=64
#SBATCH -x qhcn424
export OMPI_MCA_btl_openib_allow_ib=1
source  /home/paratera_wx_group/pwx0406/data/soft/new/moose/env.sh
export MOOSE_DIR=/home/paratera_wx_group/pwx0406/data/soft/new/moose
mpirun -np 1024 /home/paratera_wx_group/pwx0406/data/soft/new/moose/modules/porous_flow/porous_flow-opt -i test1.i --distributed-mesh

I guess it should be PBS.

**Reply by GiudGiud** [reply]
sbatch is slurm
